responses/matchers.py

```diff
@@ -402,14 +403,27 @@ def header_matcher(
     :return: (func) matcher
     """
 
+    def __compare_with_regex(request_headers: Union[Dict[Any, Any], Any]) -> bool:
+        checked_request_headers = {k: 1 for k in request_headers.keys()}
+        for k, v in headers.items():
+            if request_headers.get(k) is not None:
+                checked_request_headers[k] = 0
+                if isinstance(v, re.Pattern):
+                    if re.match(v, request_headers[k]) is None:
+                        return False
+                else:
+                    if not v == request_headers[k]:
+                        return False
+        return True if not strict_match else sum(checked_request_headers.values()) == 0
```

Comment:
can it be refactored as?
```suggestion
        for k, v in headers.items():
            if request_headers.get(k) is not None:
                if isinstance(v, re.Pattern):
                    if re.match(v, request_headers[k]) is None:
                        return False
                else:
                    if not v == request_headers[k]:
                        return False
            elif strict_match:
                return False

```

Reply:
This will work only when there's a field in `header` which doesn't match anything in `request_header`. If we have a field in `request_header` that is not present in `header`, this suggestion won't capture that.

Although this is a good optimisation, so I think merging these two like following would be perfect.

I'll also add a test for that case.

```
def _compare_with_regex(request_headers: Union[Dict[Any, Any], Any]) -> bool:
    checked_request_headers = {k: 1 for k in request_headers.keys()}
    for k, v in headers.items():
        if request_headers.get(k) is not None:
            checked_request_headers[k] = 0
            if isinstance(v, re.Pattern):
                if re.match(v, request_headers[k]) is None:
                    return False
            else:
                if not v == request_headers[k]:
                    return False
        elif strict_match:
            return False
    return True if not strict_match else sum(checked_request_headers.values()) == 0
```

Reply:
This will work only when there's a field in `header` which doesn't match anything in `request_header`.

If we have a field in `request_header` that is not present in `header`, this change won't capture that.

Although this is a good optimisation, so I think merging these two like following would be perfect.

```suggestion
        checked_request_headers = {k: 1 for k in request_headers.keys()}
        for k, v in headers.items():
            if request_headers.get(k) is not None:
                checked_request_headers[k] = 0
                if isinstance(v, re.Pattern):
                    if re.match(v, request_headers[k]) is None:
                        return False
                else:
                    if not v == request_headers[k]:
                        return False
            elif strict_match:
                return False
        return True if not strict_match else sum(checked_request_headers.values()) == 0
```

I'll also add a test for that case.

Reply:
current implementation is not human readable. I suggest we rethink the design to remove counters

Reply:
Fair point, let me add this and all kinda tests for it.

Reply:
```
def test_request_matches_headers_regex_strict_match():
    @responses.activate
    def run():

        # <some request initialization code here>

        # Subtests
        # 1. Everything matches
        # 2. len(headers) < len(request_headers)
        # 3. len(headers) > len(request_headers)
        # 4. len(headers) = len(request_headers), one field is mismatched
        # 5. len(headers) = len(request_headers), regex mismatch

        test_request_matches_headers_regex_strict_match_1(session, url)

        test_request_matches_headers_regex_strict_match_2(session, url)

        test_request_matches_headers_regex_strict_match_3(session, url)

        test_request_matches_headers_regex_strict_match_4(session, url)

        test_request_matches_headers_regex_strict_match_5(session, url)

        # defs for each function

    run()
    assert_reset()
```

just asking. Is this a good way to write tests? All are of some variation of the same functionality, hence this. Or should I split them all up?